bionic/deriver.py

```diff
@@ -273,32 +268,32 @@ def _compute_result_group_for_dnode(self, dnode):
             state = ready_task_states.pop()
 
             # If this task is already complete, we don't need to do any work.
-            # But if this is this is the first time we've seen this task, we
-            # should should log a message .
+            # But if this is the first time we've seen this task, we should
+            # should log a message.
             if state.is_complete:
                 for task_key in state.task.keys:
                     task_key_logger.log_accessed_from_memory(task_key)
                 continue
 
-            # If blocked, let's mark it and try to derive its parents.
-            if state.is_blocked():
-                ready_task_states.extend(state.parents)
-                blocked_task_key_tuples.add(state.task.keys)
+            # If blocked, let's mark it and try to derive its dependencies.
+            incomplete_dep_states = state.incomplete_dep_states()
+            if incomplete_dep_states:
+                blockage_tracker.add_blockage(
+                    blocked_state=state, blocking_states=incomplete_dep_states,
+                )
+                ready_task_states.extend(incomplete_dep_states)
                 continue
 
             # If the task isn't complete or blocked, we can complete the task.
             self._complete_task_state(state, task_key_logger)
 
-            # See if we can unblock its children after completing task state.
-            for child_state in state.children:
-                if (
-                    child_state.task.keys in blocked_task_key_tuples
-                    and not child_state.is_blocked()
-                ):
-                    ready_task_states.append(child_state)
-                    blocked_task_key_tuples.remove(child_state.task.keys)
+            # See if we can unblock any other states now that we've completed this one.
+            unblocked_states = blockage_tracker.get_unblocked_by(state)
+            ready_task_states.extend(unblocked_states)
+
+        blocked_states = blockage_tracker.get_all_blocked_states()
+        assert not blocked_states, blocked_states
```

Comment:
I don't expect you to change this since `if not seq` is in [PEP 8 guide](https://www.python.org/dev/peps/pep-0008/) but I find it hard to read.

Reply:
Yeah, this is another point I go back and forth on. I'm happy to change it to `len(blocked_states) == 0`.

Reply:
I think it's fine this way so we are consistent with the style guide, for later when someone else starts contributing to Bionic.